src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    static class Node<E> {
+        private Node<E> prev;
+        private E value;
+        private Node<E> next;
+
+        public Node(Node<E> prev, E value, Node<E> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private void checkIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException(
+                    "Index must be: 0 <= index < size for size = " + size
+            );
+        }
+    }
+
+    private Node<T> getNodeAtIndex(int index) {
+        checkIndex(index);
+        Node<T> currentNode = head;
+        int currentIndex = 0;
+        while (currentIndex < index) {
+            currentNode = currentNode.next;
+            currentIndex++;
+        }
+        return currentNode;
+    }
+
+    private void unlink(Node<T> node) {
+        node.next.prev = node.prev;
+        node.prev.next = node.next;
+    }
+
+    private void addToHead(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        head.prev = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private void addToMiddle(T value, int index) {
+        Node<T> nodeAtIndex = getNodeAtIndex(index - 1);
+        Node<T> insertNode = new Node<>(nodeAtIndex, value, nodeAtIndex.next);
+        nodeAtIndex.next.prev = insertNode;
+        nodeAtIndex.next = insertNode;
+        size++;
+    }
+
     @Override
     public void add(T value) {
+        if (!isEmpty()) {
+            Node<T> tempNext = new Node<>(tail, value, null);
+            tail.next = tempNext;
+            tail = tempNext;
+            size++;
+        }
+        if (isEmpty()) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            size++;
+        }
```

Comment:
this one is also optional, but instead of the second "if" here it's better to use "else"